File: shared/src/commonMain/kotlin/it/iterapp/core/api/AppFailure.kt
package it.iterapp.core.api

import it.iterapp.core.wire.ApiErrorCodes

/**
 * The single, platform-neutral classification of a failed operation (ADR 0015).
 * Repositories still throw [IterApiException] / [IterTransportException]; the
 * UI layer maps the caught throwable to exactly one of these via [toAppFailure]
 * so both shells render the same category the same way, and so "no connection"
 * (offline zones may still work) is never confused with "server unreachable".
 */
sealed interface AppFailure {

  /** The device has no network path at all. Downloaded offline zones still work. */
  data object NoConnection : AppFailure

  /** Network is up, but the gateway can't be reached (DNS/refused/reset). */
  data class ServerUnreachable(val code: String? = null) : AppFailure

  /** The request took too long (a client or server timeout fired). */
  data object Timeout : AppFailure

  /** The gateway asked us to slow down (HTTP 429 / `BUSY`). */
  data class RateLimited(val retryAfterSeconds: Int? = null) : AppFailure

  /** A non-swallowed 404 — the resource genuinely does not exist. */
  data object NotFound : AppFailure

  /** The request was rejected as invalid (4xx with a machine [code]). */
  data class BadRequest(val code: String? = null) : AppFailure

  /** The gateway returned a server-side fault envelope (5xx / upstream / GraphQL). */
  data class Server(val code: String? = null) : AppFailure

  /** An offline-bundle operation failed (area too large, storage full, corrupt). */
  data class OfflineBundle(val code: String? = null) : AppFailure

  /** Anything the client could not classify. */
  data object Unknown : AppFailure

  /** Stable id for logging / cross-language mapping; never localized. */
  val id: String
    get() = when (this) {
      NoConnection -> "no_connection"
      is ServerUnreachable -> "server_unreachable"
      Timeout -> "timeout"
      is RateLimited -> "rate_limited"
      NotFound -> "not_found"
      is BadRequest -> "bad_request"
      is Server -> "server"
      is OfflineBundle -> "offline_bundle"
      Unknown -> "unknown"
    }
}

/** Offline-surface codes that mean "the bundle request itself was the problem". */
private val OFFLINE_BUNDLE_CODES = setOf(
  ApiErrorCodes.AREA_TOO_LARGE,
  ApiErrorCodes.PAYLOAD_TOO_LARGE,
  ApiErrorCodes.STORE_FULL,
  ApiErrorCodes.EXTRACT_FAILED,
  ApiErrorCodes.BBOX_REQUIRED,
  ApiErrorCodes.BBOX_INVALID,
  ApiErrorCodes.BBOX_OUT_OF_RANGE,
  ApiErrorCodes.BBOX_DEGENERATE,
  ApiErrorCodes.ZOOM_INVALID,
)

/** Codes that mean the gateway reached a broken/absent upstream. */
private val UPSTREAM_CODES = setOf(
  ApiErrorCodes.UPSTREAM_UNAVAILABLE,
  ApiErrorCodes.UPSTREAM_ERROR,
  ApiErrorCodes.INTERNAL,
)

/**
 * Maps a caught throwable to its [AppFailure]. [isOnline] splits a bare
 * transport failure into [AppFailure.NoConnection] vs
 * [AppFailure.ServerUnreachable]; pass the current connectivity reading.
 */
fun Throwable.toAppFailure(isOnline: Boolean = true): AppFailure = when (this) {
  is IterTransportException -> when (kind) {
    TransportKind.TIMEOUT -> AppFailure.Timeout
    TransportKind.UNREACHABLE ->
      if (isOnline) AppFailure.ServerUnreachable() else AppFailure.NoConnection
  }

  is IterApiException -> classifyApi(status, code)

  else -> AppFailure.Unknown
}

private fun classifyApi(status: Int, code: String?): AppFailure = when {
  code != null && code in OFFLINE_BUNDLE_CODES -> AppFailure.OfflineBundle(code)
  status == 429 || code == ApiErrorCodes.BUSY -> AppFailure.RateLimited()
  status == 404 -> AppFailure.NotFound
  status in 500..599 || (code != null && code in UPSTREAM_CODES) -> AppFailure.Server(code)
  code == ApiErrorCodes.TIMEOUT -> AppFailure.Timeout
  // GraphQL body errors are synthesized as IterApiException(200, code, …).
  status == 200 && code != null -> AppFailure.Server(code)
  status in 400..499 -> AppFailure.BadRequest(code)
  else -> AppFailure.Unknown
}

File: iosApp/Sources/Support/AppFailure.swift
import IterCore
import SwiftUI

/// Swift twin of the shared `AppFailure` taxonomy (ADR 0015). Repositories
/// still throw `IterApiException` / `IterTransportException`; the shell catches
/// and classifies once, so every surface renders the same category the same
/// way and "no connection" (offline zones may still work) is never confused
/// with "server unreachable".
enum AppFailure: Equatable {
  case noConnection
  case serverUnreachable(code: String?)
  case timeout
  case rateLimited
  case notFound
  case badRequest(code: String?)
  case server(code: String?)
  case offlineBundle(code: String?)
  case unknown

  /// Offline-surface codes that mean the bundle request itself was the problem.
  private static let offlineBundleCodes: Set<String> = [
    "AREA_TOO_LARGE", "PAYLOAD_TOO_LARGE", "STORE_FULL", "EXTRACT_FAILED",
    "BBOX_REQUIRED", "BBOX_INVALID", "BBOX_OUT_OF_RANGE", "BBOX_DEGENERATE",
    "ZOOM_INVALID",
  ]

  /// Codes that mean the gateway reached a broken/absent upstream.
  private static let upstreamCodes: Set<String> = [
    "UPSTREAM_UNAVAILABLE", "UPSTREAM_ERROR", "INTERNAL",
  ]

  /// Maps a caught error to its `AppFailure`. Mirrors the shared
  /// `Throwable.toAppFailure` / `classifyApi` exactly. A Kotlin exception
  /// crossing the `@Throws` boundary arrives as an `NSError` carrying the
  /// original throwable under the "KotlinException" userInfo key.
  static func from(_ error: Error, isOnline: Bool) -> AppFailure {
    let kotlinException = (error as NSError).userInfo["KotlinException"]

    if let transport = kotlinException as? IterTransportException {
      switch transport.kind {
      case .timeout:
        return .timeout
      case .unreachable:
        return isOnline ? .serverUnreachable(code: nil) : .noConnection
      default:
        return isOnline ? .serverUnreachable(code: nil) : .noConnection
      }
    }

    if let api = kotlinException as? IterApiException {
      return classifyApi(status: Int(api.status), code: api.code)
    }

    return .unknown
  }

  private static func classifyApi(status: Int, code: String?) -> AppFailure {
    if let code, offlineBundleCodes.contains(code) {
      return .offlineBundle(code: code)
    }
    if status == 429 || code == "BUSY" {
      return .rateLimited
    }
    if status == 404 {
      return .notFound
    }
    if (500...599).contains(status) || (code.map { upstreamCodes.contains($0) } ?? false) {
      return .server(code: code)
    }
    if code == "TIMEOUT" {
      return .timeout
    }
    // GraphQL body errors are synthesized as IterApiException(200, code, …).
    if status == 200, code != nil {
      return .server(code: code)
    }
    if (400...499).contains(status) {
      return .badRequest(code: code)
    }
    return .unknown
  }
}

/// The reusable "fancy" error surface, fed by an `AppFailure`. Renders as a
/// system `ContentUnavailableView` (icon + title + message + optional Retry).
/// `hasOfflineMaps` tailors the no-connection copy; `onRetry` renders a Retry
/// button when the failure is recoverable.
struct StatusView: View {
  let failure: AppFailure
  var hasOfflineMaps = false
  var onRetry: (() -> Void)?

  private var symbol: String {
    switch failure {
    case .noConnection: return "wifi.slash"
    case .serverUnreachable, .server: return "exclamationmark.icloud"
    case .timeout: return "clock.arrow.circlepath"
    case .rateLimited: return "hourglass"
    case .notFound: return "magnifyingglass"
    case .offlineBundle: return "arrow.down.circle.badge.xmark"
    case .badRequest, .unknown: return "exclamationmark.triangle"
    }
  }

  private var title: String {
    switch failure {
    case .noConnection: return Strings.errorOfflineTitle
    case .serverUnreachable, .server: return Strings.errorUnreachableTitle
    case .timeout: return Strings.errorTimeoutTitle
    case .rateLimited: return Strings.errorBusyTitle
    case .notFound: return Strings.errorNotFoundTitle
    case .offlineBundle(let code): return offlineBundleTitle(code)
    case .badRequest, .unknown: return Strings.errorGenericTitle
    }
  }

  private var message: String {
    switch failure {
    case .noConnection:
      return hasOfflineMaps ? Strings.errorOfflineBodyWithMaps : Strings.errorOfflineBody
    case .serverUnreachable, .server: return Strings.errorUnreachableBody
    case .timeout: return Strings.errorTimeoutBody
    case .rateLimited: return Strings.errorBusyBody
    case .notFound: return Strings.errorNotFoundBody
    case .offlineBundle(let code): return offlineBundleMessage(code)
    case .badRequest, .unknown: return Strings.errorGenericBody
    }
  }

  /// A retryable failure gets a Retry button; a bad request / not-found does not.
  private var showsRetry: Bool {
    guard onRetry != nil else { return false }
    switch failure {
    case .badRequest, .notFound: return false
    default: return true
    }
  }

  private func offlineBundleTitle(_ code: String?) -> String {
    switch code {
    case "AREA_TOO_LARGE", "PAYLOAD_TOO_LARGE": return Strings.offlineErrorTooLarge
    case "STORE_FULL": return Strings.errorGenericTitle
    default: return Strings.offlineErrorInvalidArea
    }
  }

  private func offlineBundleMessage(_ code: String?) -> String {
    switch code {
    case "AREA_TOO_LARGE", "PAYLOAD_TOO_LARGE": return Strings.offlineErrorTooLarge
    case "STORE_FULL": return Strings.errorGenericBody
    default: return Strings.offlineErrorInvalidArea
    }
  }

  var body: some View {
    // Deployment target is iOS 17 (project.yml), so ContentUnavailableView is
    // always available — the app's "fancy", system-native empty/error state.
    ContentUnavailableView {
      Label(title, systemImage: symbol)
    } description: {
      Text(message)
    } actions: {
      if showsRetry, let onRetry {
        Button(Strings.actionRetry, action: onRetry)
          .buttonStyle(.borderedProminent)
      }
    }
  }
}
